### PR TITLE
Update the value of f when the network is open

### PR DESCRIPTION
**Pull request by ashamis**

### CODE REVIEW

**Comment by cimetrics** [comment]
images

File: ePBFT/src/pbft/libbyz/Node.cpp [hunk 1]
```diff
@@ -368,3 +368,10 @@ void Node::send_to_replicas(Message* m)
     }
   }
 }
+
+void Node::set_f(uint32_t f)
```
Comment (olgavrou) [inline ePBFT/src/pbft/libbyz/Node.cpp]: Can f be a size_t or is there a reason we want it to be uint32_t?

File: src/kv/kvtypes.h [hunk 2]
```diff
@@ -191,6 +191,8 @@ namespace kv
     virtual void enable_all_domains() {}
     virtual void resume_replication() {}
     virtual void suspend_replication(kv::Version) {}
+
+    virtual void set_f(uint32_t f) = 0;
```
Comment (olgavrou) [inline src/kv/kvtypes.h]: It would be better to move the basic implementation from `raftconsensus` here so that `raftconsensus` doesn't need to worry about it, same as `on_request`.

File: src/node/nodestate.h [hunk 3]
```diff
@@ -1081,6 +1081,8 @@ namespace ccf
                                         const Service::Write& w) {
         if (w.at(0).value.status == ServiceStatus::OPEN)
         {
+          this->consensus->set_f(
+            1); // TODO: we should make f come from a KV table
```
Comment (olgavrou) [inline src/node/nodestate.h]: Maybe ticket this?

File: tests/e2e_logging_pbft.py [hunk 4]
```diff
@@ -34,36 +34,39 @@ def run(args):
...
+        # network.open_network(primary)
+        script = None
+        result = network.propose(1, primary, script, None, "open_network")
+        network.vote_using_majority(primary, result[1]["id"], False)
         LOG.info("***** Network is now open *****")
 
-        with primary.node_client() as mc:
-            check_commit = infra.ccf.Checker(mc)
-            check = infra.ccf.Checker()
+        # with primary.node_client() as mc:
```
Comment (olgavrou) [inline tests/e2e_logging_pbft.py]: has this been left commented out by mistake?

File: src/kv/kvtypes.h [hunk 5]
```diff
@@ -191,6 +191,8 @@ namespace kv
     virtual void enable_all_domains() {}
     virtual void resume_replication() {}
     virtual void suspend_replication(kv::Version) {}
```
Comment (olgavrou) [inline src/kv/kvtypes.h]: Maybe define `f`'s type in `kvtypes.h` same as `View` and `Seqno`. Could also maybe be defined in `types.h` in ePBFT. Unless it does end up switching to `size_t`

File: src/kv/kvtypes.h (hunk 2 above)
Comment (achamayou) [reply src/kv/kvtypes.h]: We don't want any implementation in kvtypes (or in anything *types in general), I agree it's probably good to share this logic, but not here.

File: src/node/nodestate.h (hunk 3 above)
Comment (achamayou) [reply src/node/nodestate.h]: This isn't so clear cut, f could also be ephemeral and derived by each node when it hits network open (that is in a table), based on what it has in its nodes table/list.

File: ePBFT/src/pbft/libbyz/Certificate.h [hunk 6]
```diff
@@ -176,6 +176,8 @@ class Certificate
   T* mym; // my message in this or null if I have no message in this
   Time t_sent; // time at which mym was last sent
 
+  uint32_t f; // the value of f when starting to run
```
Comment (jumaffre) [inline ePBFT/src/pbft/libbyz/Certificate.h]: I would add a comment here along these lines: "Note that is is necessary that f stays constant for the lifetime of the certificate since a request may change the value of f".

File: ePBFT/src/pbft/libbyz/New_key.cpp [hunk 7]
```diff
@@ -79,6 +79,22 @@ bool New_key::verify()
   {
     // Received message from unknown sender
     LOG_INFO << "Request from unknown pricipal, id:" << id() << std::endl;
+
+    std::string pubk_sig =
```
Comment (jumaffre) [inline ePBFT/src/pbft/libbyz/New_key.cpp]: Why is this necessary?

File: src/kv/kvtypes.h (hunk 2 above)
Comment (olgavrou) [reply src/kv/kvtypes.h]: It's the same as `on_request`, we dummy implement it in the Consensus class so that when Raft extends this method there is no need for it to be overridden.

File: ePBFT/src/pbft/libbyz/Replica.cpp [hunk 8]
```diff
@@ -1719,6 +1719,11 @@ void Replica::execute_prepared(bool committed)
         }
       }
     }
+
+    if (global_commit_cb != nullptr)
```
Comment (jumaffre) [inline ePBFT/src/pbft/libbyz/Replica.cpp]: Why is this necessary?

File: tests/e2e_logging_pbft.py (hunk 4 above)
Comment (ashamis) [reply tests/e2e_logging_pbft.py]: Eventually the comments will be removed and we will be able to run this code, but we are not there yet

File: ePBFT/src/pbft/libbyz/New_key.cpp (hunk 7 above)
Comment (ashamis) [reply ePBFT/src/pbft/libbyz/New_key.cpp]: This is a layover from stand-alone pbft. It will not be needed when we have the integration fully complete and can test ePBFT just inside of CCF

File: ePBFT/src/pbft/libbyz/Replica.cpp (hunk 8 above)
Comment (ashamis) [reply ePBFT/src/pbft/libbyz/Replica.cpp]: You want to send the commits for the message that opens the network before it is open as opening will mean f == 1 and we will try to send this message to all the other nodes but not yourself and since this message started when f == 0 it expects that you will not send it to the other nodes

File: ePBFT/src/pbft/libbyz/Certificate.h (hunk 6 above)
Comment (ashamis) [reply ePBFT/src/pbft/libbyz/Certificate.h]: i changed it to be const which I believe implies the same thing

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

File: src/kv/kvtypes.h (hunk 2 above)
Comment (achamayou) [reply src/kv/kvtypes.h]: Let's talk about this offline, it should be split out. ktypes.h should be a pure interface.

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images
